### PR TITLE
Rewrite GithubStats in TypeScript

### PR DESCRIPTION
**Pull request by karanjitsingh**
This pr is still incomplete, will be push more commits

Addresses #7 #3 

- Introduce package.json
  - typescript
  - tslint-microsoft-config
  - tslint
- Major Todos:
  - setup local build
  - typescript compilation
  - ci pipeline @byliuyang, since we're introducing typing you might want to look into what ci solutions you want. I recommend azure devops, since i'm already familiar with it I can help set up a pipeline
  - define the rest of internal models
  - separate file for interfaces and types
  - github api model

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "github-stats",
+  "version": "1.0.0",
+  "description": "> Embed Github profile every where you want",
```
Comment (naturetime10) [inline package.json]:
```suggestion
  "description": "Embed Github profile every where you want",
```

File: package.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  "dependencies": {
+    "tslint-eslint-rules": "^5.4.0"
+  },
+  "devDependencies": {
+    "tslint": "^6.0.0",
+    "tslint-microsoft-contrib": "^6.2.0",
+    "typescript": "^3.7.5"
+  },
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (naturetime10) [inline package.json]: Would you like to open an issue for us to add unit tests? :wink:

File: package.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/karanjitsingh/github-stats.git"
```
Comment (naturetime10) [inline package.json]:
```suggestion
    "url": "https://github.com/byliuyang/github-stats.git"
```

File: package.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  },
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/karanjitsingh/github-stats/issues"
```
Comment (naturetime10) [inline package.json]:
```suggestion
    "url": "https://github.com/byliuyang/github-stats/issues"
```

File: package.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  },
+  "homepage": "https://github.com/karanjitsingh/github-stats#readme"
```
Comment (naturetime10) [inline package.json]:
```suggestion
  "homepage": "https://github.com/byliuyang/github-stats#readme"
```

File: src/github-stats.ts [hunk 6]
```diff
@@ -0,0 +1,498 @@
+// TODO github api model
```
Comment (naturetime10) [inline src/github-stats.ts]:
```suggestion
// TODO(issue#the_issue_number): define Github API entities.
```
This format is not only for Webstorm to automatically parse and highlight them in the IDE but also allowing us, as a team to track future items!

File: src/github-stats.ts [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+    repoLanguages: RepoLanugages;
+}
+
+type JsonCacheType = keyof JsonCacheOptionsMap;
+
+type Language = keyof LANGUAGE_TO_COLOR;
+
+type LanguageCountMap = { [key in Language]?: number };
+
+export async function GithubStats(username: string): Promise<GithubStats> {
```
Comment (naturetime10) [inline src/github-stats.ts]: Nit: I wonder whether we can make this an TypeScript class and abstract out all the I/O operations through interface and use constructor dependency injection to initialize this class. However, that may be complicated for the users to use.

Comment (naturetime10) [reply src/github-stats.ts]: TODO: Ideally, in the future PRs, we want to separate caching, Github API, and SVG rendering into their own class to help simplify the logic and improves testability of the code:

```typescript
class GithubStats {
    private cache: Cache;
    private githubAPI: GithubAPI; // parsing the home page into JSON object and call Github languages APIs
    private svgRender: SvgRender; // Render JSON object into diagrams on the Web
}
```

Let's together kill my legacy tech debts!

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Comment by naturetime10** [comment]
Hi @karanjitsingh, thank you so much for awesome rewrite of this library! I left few comments, please check them out.

Let open issues for all the TODOs you suggested in the PR description so that we, as a team, can address them one by one.

Let's setup Azure CI in the future PRs so that we can guard code quality and eliminate some of the potential bugs!

File: src/github-stats.ts (hunk 7 above)
Comment (karanjitsingh) [reply src/github-stats.ts]: I started with a similar idea, but it seems to be adding a lot of complexity to a simple thing. Maybe when we extend this further we can go ahead with that. For now just reorganizing and adding some typing.

**Comment by karanjitsingh** [comment]
@byliuyang, test this locally, seems to be working on my end.

build with `npm run build`

**Comment by Spiderpig86** [comment]
Just a slight refactor suggestion, but would it be more appropriate to split `models.ts` into multiple files inside a `models` directory?

**Comment by naturetime10** [comment]
@Spiderpig86 Do you want to work on model split refactor?!!!

**Comment by naturetime10** [comment]
Will review this PR soon

File: package.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "github-stats",
+  "version": "1.0.0",
+  "description": "Embed Github profile every where you want",
```
Comment (naturetime10) [inline package.json]:
```suggestion
  "description": "Embed Github profile everywhere you want",
```

**Review by naturetime10 — APPROVED** [review]

**Comment by Spiderpig86** [comment]
> @Spiderpig86 Do you want to work on model split refactor?!!!

Will work on new PR when there is time. 👍

**Pull request merged**